hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionRegistryFactory.java

```diff
@@ -17,27 +17,75 @@
  */
 package org.apache.hadoop.hbase.client;
 
-import static org.apache.hadoop.hbase.HConstants.CLIENT_CONNECTION_REGISTRY_IMPL_CONF_KEY;
-
+import java.io.IOException;
+import java.net.URI;
+import java.util.ServiceLoader;
+import org.apache.commons.lang3.StringUtils;
 import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.HConstants;
 import org.apache.hadoop.hbase.security.User;
 import org.apache.hadoop.hbase.util.ReflectionUtils;
 import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hbase.thirdparty.com.google.common.collect.ImmutableMap;
 
 /**
- * Factory class to get the instance of configured connection registry.
+ * The entry point for creating a {@link ConnectionRegistry}.
  */
 @InterfaceAudience.Private
 final class ConnectionRegistryFactory {
 
+  private static final Logger LOG = LoggerFactory.getLogger(ConnectionRegistryFactory.class);
+
+  private static final ImmutableMap<String, ConnectionRegistryCreator> CREATORS;
+  static {
+    ImmutableMap.Builder<String, ConnectionRegistryCreator> builder = ImmutableMap.builder();
+    for (ConnectionRegistryCreator factory : ServiceLoader.load(ConnectionRegistryCreator.class)) {
+      builder.put(factory.protocol(), factory);
```

Comment:
Wikipedia says that schemes should be case-insensitive. We should normalize these before adding them to the map.